lib/block-supports/settings.php

```diff
@@ -0,0 +1,119 @@
+	if ( ! $block_content ) {
+		return $block_content;
+	}
+
+	// return early if the block doesn't have support for settings.
+	$block_type           = WP_Block_Type_Registry::get_instance()->get_registered( $block['blockName'] );
+	if ( ! block_has_support( $block_type, array( '__experimentalSettings' ), false ) ) {
+		return $block_content;
+	}
+
+	// return early if no settings are found on the block attributes.
+	$block_settings = _wp_array_get( $block, array( 'attrs', 'settings'), null );
+	if ( empty( $block_settings ) ) {
+		return $block_content;
+	}
+
+	$class_name = _gutenberg_get_presets_class_name( $block );
+
+	// Like the layout hook this assumes the hook only applies to blocks with a single wrapper.
+	// Retrieve the opening tag of the first HTML element.
+	$html_element_matches = array();
+	preg_match( '/<[^>]+>/', $block_content, $html_element_matches, PREG_OFFSET_CAPTURE );
+	$first_element = $html_element_matches[0][0];
+	// If the first HTML element has a class attribute just add the new class
+	// as we do on layout and duotone.
+	if ( strpos( $first_element, 'class="' ) !== false ) {
+		$content = preg_replace(
+			'/' . preg_quote( 'class="', '/' ) . '/',
+			'class="' . $class_name . ' ',
+			$block_content,
+			1
+		);
+	} else {
+		// If the first HTML element has no class attribute we should inject the attribute before the attribute at the end.
+		$first_element_offset = $html_element_matches[0][1];
+		$content              = substr_replace( $block_content, ' class="' . $class_name . '"', $first_element_offset + strlen( $first_element ) - 1, 0 );
+	}
+
+	return $content;
+}
+
+/**
+ * Render the block level presets stylesheet.
+ *
+ * @access private
+ *
+ * @param string|null $pre_render   The pre-rendered content. Default null.
+ * @param array       $block The block being rendered.
+ *
+ * @return null
+ */
+function _gutenberg_add_block_level_preset_styles( $pre_render, $block ) {
+	// Return early if the block has not support for descendent block styles.
+	$block_type           = WP_Block_Type_Registry::get_instance()->get_registered( $block['blockName'] );
+	if ( ! block_has_support( $block_type, array( '__experimentalSettings' ), false ) ) {
+		return null;
+	}
+
+	// return early if no settings are found on the block attributes.
+	$block_settings = _wp_array_get( $block, array( 'attrs', 'settings' ), null );
+	if ( empty( $block_settings ) ) {
+		return null;
+	}
+
+	$class_name = '.' . _gutenberg_get_presets_class_name( $block );
+
+	// Remove any potentially unsafe styles.
+	$theme_json_shape = WP_Theme_JSON_Gutenberg::remove_insecure_properties( array(
+		'version' =>  WP_Theme_JSON_Gutenberg::LATEST_SCHEMA,
+		'settings' => $block_settings,
+	) );
+	$theme_json_object = new WP_Theme_JSON_Gutenberg( $theme_json_shape );
+
+	$styles = '';
+
+	// include preset css variables declaration on the stylesheet.
+	$styles .= $theme_json_object->get_scoped_css_variables( $class_name );
```

Comment:
We are doing a few things in this method (the same holds for `get_scoped_css_classes`):

- provide a new root selector
- scope the existing block selectors with the provided class

This logic can't be detached from the consumer that uses it (rendering a stylesheet for _block settings_), which makes me wonder if this code should live here at all.

How could we improve this? :thinking:

Reply:
I was expecting we could just do:

```php
$theme_json = new WP_Theme_JSON( $settings );
$theme_json->get_stylesheet();
```

and find a way to provide a new root selector and the class to scope the styles with.

It seems that `get_stylesheet` is already tailored for a single use case: the "global styles stylesheet". While the code has remained agnostic of context (it's a function of the provided theme.json data and the block registry) its output contains [layout data](https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-includes/class-wp-theme-json.php#L879) that is only relevant in a "global stylesheet".

Reply:
In that case, we would need to add the ability to scope selectors to the get_stylesheet function and to provide a root selector.
Eg:
```
$theme_json = new WP_Theme_JSON( $settings );
$theme_json->get_stylesheet( $types = array( 'variables', 'styles', 'presets' ), $origins = null, $root_selector, $scope);
```

Would something like this work?

Reply:
Can't think of a better alternative. Others I've considered:

- Creating a `WP_Theme_JSON_Section` class that overwrites the bits of `WP_Theme_JSON` that are "global styles" specific.
- Creating a new `get_stylesheet_scoped` method.

All of them are worse than expanding `get_stylesheet`. For this use case (`variables` and `presets`), we should be able to make the existing `get_stylesheet` work. If the time comes to also render `styles` we'd need to make further modifications (understand how layout data should be rendered in different contexts, etc.).

Perhaps we can expand `get_stylesheet` with a new `$meta` array parameter to configure this behavior so we have leeway to add more if the future, should we need.